ipa-core/src/query/runner/hybrid.rs

```diff
@@ -1,37 +1,285 @@
-use std::{marker::PhantomData, sync::Arc};
+use std::{
+    collections::HashSet,
+    marker::PhantomData,
+    sync::{Arc, Mutex},
+};
+
+use futures::{stream::iter, StreamExt, TryStreamExt};
 
 use crate::{
     error::Error,
+    ff::boolean_array::{BA20, BA3, BA8},
     helpers::{
         query::{HybridQueryParams, QuerySize},
-        BodyStream,
+        BodyStream, LengthDelimitedStream,
     },
     hpke::PrivateKeyRegistry,
+    protocol::{context::UpgradableContext, ipa_prf::shuffle::Shuffle, step::ProtocolStep::Hybrid},
+    report::{hybrid::HybridReport, EncryptedOprfReport},
     secret_sharing::{replicated::semi_honest::AdditiveShare as ReplicatedShare, SharedValue},
 };
 
 pub struct Query<C, HV, R: PrivateKeyRegistry> {
-    _config: HybridQueryParams,
-    _key_registry: Arc<R>,
+    config: HybridQueryParams,
+    key_registry: Arc<R>,
     phantom_data: PhantomData<(C, HV)>,
 }
 
-impl<C, HV: SharedValue, R: PrivateKeyRegistry> Query<C, HV, R> {
+impl<C, HV: SharedValue, R: PrivateKeyRegistry> Query<C, HV, R>
+where
+    C: UpgradableContext + Shuffle,
+{
     pub fn new(query_params: HybridQueryParams, key_registry: Arc<R>) -> Self {
         Self {
-            _config: query_params,
-            _key_registry: key_registry,
+            config: query_params,
+            key_registry,
             phantom_data: PhantomData,
         }
     }
 
     #[tracing::instrument("hybrid_query", skip_all, fields(sz=%query_size))]
     pub async fn execute(
         self,
-        _ctx: C,
+        ctx: C,
         query_size: QuerySize,
-        _input_stream: BodyStream,
+        input_stream: BodyStream,
     ) -> Result<Vec<ReplicatedShare<HV>>, Error> {
-        unimplemented!()
+        let Self {
+            config,
+            key_registry,
+            phantom_data: _,
+        } = self;
+        tracing::info!("New hybrid query: {config:?}");
+        let _ctx = ctx.narrow(&Hybrid);
+        let sz = usize::from(query_size);
+
+        let seen_encrypted_reports: Arc<Mutex<HashSet<Vec<u8>>>> =
+            Arc::new(Mutex::new(HashSet::with_capacity(sz)));
+
+        let _input = if config.plaintext_match_keys {
+            unimplemented!()
+        } else {
+            LengthDelimitedStream::<EncryptedOprfReport<BA8, BA3, BA20, _>, _>::new(input_stream)
+                .map_err(Into::<Error>::into)
+                .map_ok(move |enc_reports| {
+                    iter(enc_reports.into_iter().map({
+                        let seen_encrypted_reports =
+                            Arc::<Mutex<HashSet<Vec<u8>>>>::clone(&seen_encrypted_reports);
+
+                        let key_registry = Arc::<R>::clone(&key_registry);
+                        move |enc_report| {
+                            let ciphertext = enc_report.mk_ciphertext().to_owned();
+                            let mut guard = seen_encrypted_reports.lock().unwrap();
```

Comment:
I think `Arc<Mutex` is unnecessary here. Stream is polled by one thread only, so we don't need synchronization here. Rust probably steered you towards that path because of double closure usage here, but it is not really needed. Something like this should work (and if you create a struct, this would be even better)

```rust
        let mut seen_encrypted_reports: HashSet<Vec<u8>> = HashSet::with_capacity(sz);
        ...
        LengthDelimitedStream::<EncryptedOprfReport<BA8, BA3, BA20, _>, _>::new(input_stream)
                .map_err(Into::<Error>::into)
                .map_ok(|enc_reports| {
                    // check uniqueness
                    let before = seen_encrypted_reports.len();
                    // extend, or iterate over `enc_reports` - both should work
                    seen_encrypted_reports.extend(enc_reports.iter().map(|report| report.mk_ciphertext().to_vec()));
                    assert_eq!(before + enc_reports.len(), seen_encrypted_reports.len());
                    
                    iter(enc_reports.into_iter().map(|enc_report|
                        HybridReport::<BA8, BA3>::from_encrypted_oprf_report::<R, _, BA20>(
                            &enc_report,
                            &key_registry,
                        ).map_err(Into::<Error>::into)))
                })
```

Reply:
is this doing some sort of automatic chunking? (just trying to reason about what's different here)

Rust steered me towards `Arc<Mutux` because it's async, I believe. but let me see if it complains about this (and also the struct approach.)

Reply:
it is all about returning an iterator that grabs a closure I think. async can and should work with references only here - there is only one thread polling this stream.

I tested the code above and it worked for me, although I am not entirely happy using `extend` here